fe/src/main/java/org/apache/doris/alter/RollupJobV2.java

```diff
@@ -0,0 +1,744 @@
+import org.apache.doris.task.AgentTaskQueue;
+import org.apache.doris.task.AlterReplicaTask;
+import org.apache.doris.task.CreateReplicaTask;
+import org.apache.doris.thrift.TStorageMedium;
+import org.apache.doris.thrift.TStorageType;
+import org.apache.doris.thrift.TTaskType;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.TimeUnit;
+
+/*
+ * Author: Chenmingyu
+ * Date: Jul 8, 2019
+ */
+
+/*
+ * Version 2 of RollupJob.
+ * This is for replacing the old RollupJob
+ * https://github.com/apache/incubator-doris/issues/1429
+ */
+public class RollupJobV2 extends AlterJobV2 {
+    private static final Logger LOG = LogManager.getLogger(RollupJobV2.class);
+
+    // partition id -> (rollup tablet id -> base tablet id)
+    private Map<Long, Map<Long, Long>> partitionIdToBaseRollupTabletIdMap = Maps.newHashMap();
+    private Map<Long, MaterializedIndex> partitionIdToRollupIndex = Maps.newHashMap();
+
+    // rollup and base schema info
+    private long baseIndexId;
+    private long rollupIndexId;
+    private String baseIndexName;
+    private String rollupIndexName;
+
+    private List<Column> rollupSchema = Lists.newArrayList();
+    private int baseSchemaHash;
+    private int rollupSchemaHash;
+
+    private KeysType rollupKeysType;
+    private short rollupShortKeyColumnCount;
+
+    // The rollup job will wait all transactions before this txn id finished, then send the rollup tasks.
+    protected long watershedTxnId = -1;
+
+    // save all create rollup tasks
+    private AgentBatchTask rollupBatchTask = new AgentBatchTask();
+
+    public RollupJobV2(long jobId, long dbId, long tableId, String tableName, long timeoutMs,
+            long baseIndexId, long rollupIndexId, String baseIndexName, String rollupIndexName,
+            List<Column> rollupSchema, int baseSchemaHash, int rollupSchemaHash,
+            KeysType rollupKeysType, short rollupShortKeyColumnCount) {
+        super(jobId, JobType.ROLLUP, dbId, tableId, tableName, timeoutMs);
+
+        this.baseIndexId = baseIndexId;
+        this.rollupIndexId = rollupIndexId;
+        this.baseIndexName = baseIndexName;
+        this.rollupIndexName = rollupIndexName;
+
+        this.rollupSchema = rollupSchema;
+        this.baseSchemaHash = baseSchemaHash;
+        this.rollupSchemaHash = rollupSchemaHash;
+        this.rollupKeysType = rollupKeysType;
+        this.rollupShortKeyColumnCount = rollupShortKeyColumnCount;
+    }
+
+    private RollupJobV2() {
+        super(JobType.ROLLUP);
+    }
+
+    public void addTabletIdMap(long partitionId, long rollupTabletId, long baseTabletId) {
```

Comment:
does it need thread safe?

Reply:
No need, it only be accessed when job is creating.